### PR TITLE
fix(space): post-approval reuses live agent session, else creates (#850, #816)

### PR DESCRIPTION
**Pull request by lsm**
Fixes the 100% post-approval merge failure (#816/#2323) and simplifies the design: approval is a task-level event, not tied to the node that submitted.

Two layers:

1. **Per-delivery (revert of the force-fresh workaround):** `spawnPostApprovalSubSession` reuses the target agent's live session if it has one, else creates one — and injects the kickoff directly (never interrupting). The earlier `forceNewSession` flag worked but for the wrong reason; the real #816 cause was `createSubSession`'s reuse path calling `restartQuery()` (`reinjectNodeAgentMcpServer`), which interrupts an active drive. Direct injection never interrupts.

2. **Fan-out (this PR's design change):** `PostApprovalRouter` now scans every node carrying a `postApproval` route and dispatches each, independent of the source node. The built-in Coding/Research/QA workflows move their merge route off the Review/QA end node onto the dedicated Post-Approval (merger) node — the node whose own agent runs the merge. Existing spaces migrate for free via `mergeNodeStructuralFieldsFromTemplate` (overwrites node postApproval from the template by name).

Also renames the misleading "Retry" button to "Send back" in PendingPostApprovalBanner (resets approved→in_progress; doesn't re-run the merge).

Follow-ups (deferred, non-blocking): rehydration of the fresh merger session on daemon restart; flush-pollution on the create path; the agent-level field move (`postApproval`: node → agent slot) as a pure structural refactor.

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (GLM)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** GLM

Recommendation: APPROVE

Verified the root cause, the fix, and the hard constraints. Zero findings.

**Root cause confirmed.** `PostApprovalRouter.route` calls `spawnPostApprovalSubSession` at `post-approval-router.ts:408` and only stamps `postApprovalSessionId` at `:420`. With the old unconditional agentName reuse, `createSubSession` returned the reviewer's still-live session, the merge kickoff was injected into a session the approval/quiesce flow was interrupting, and the SDK threw "Interrupted by user" — so the spawn threw *before* `:420` and the task parked in `approved`. Matches the #816/#2323 repro exactly.

**Fix is correct and surgical.** `SubSessionMemberInfo.forceNewSession` skips the reuse lookup (`if (memberInfo?.agentName && !memberInfo.forceNewSession)`, `task-agent-manager.ts:1280`) and is set only from `spawnPostApprovalSubSession` (`:4859`). Grep confirms exactly three production references — the flag cannot leak into any other caller, so normal node re-execution (a second review cycle) still reuses. Hard constraint satisfied.

**No completion-callback regression.** The fresh path in `createSubSession` does not register an idle-completion callback (unlike the reuse path at `:1395`), but post-approval completion does not use it: `mark_complete` (`end-node-handlers.ts:189`) directly transitions `approved → done` via `setTaskStatus`. The idle callback is for normal node-turn completion, so its absence here is correct, not a gap. The sibling-quiesce exclusion (`space-runtime.ts:8014`) protects the freshly-spawned id.

**Test is adequate.** The new test exercises the real reuse-vs-fresh *decision* on a real `TaskAgentManager` (only `AgentSession.fromInit` and the MCP-injection helpers are stubbed), asserting both (a) `forceNewSession` yields a fresh id distinct from the reviewer's with `fromInit` invoked, and (b) without the flag the second activation reuses the prior session with `fromInit` not invoked. Framework is correct — `bun:test` is aliased to a vitest shim (`vitest.config.ts:32-35`), consistent with all 445 sibling daemon suites.

**Verification run locally:** typecheck clean, oxlint clean, 31 post-approval tests pass (router + dispatch-integration + fresh-session), full `5-space/agent` shard green.

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts [hunk 1]
```diff
@@ -4834,6 +4850,13 @@ export class TaskAgentManager {
       agentId: matchedSlot.agentId,
       agentName: matchedSlot.name,
       nodeId: matchedNodeId,
+      // Post-approval MUST spawn a fresh session. The target agent (e.g.
+      // 'reviewer') already has a live session from the review cycle; reusing
+      // it injects the merge kickoff into a session the approval/quiesce flow
+      // is interrupting, so the SDK throws "Interrupted by user" before the
+      // router can stamp postApprovalSessionId — parking the task in
+      // 'approved' with the merge never running.
+      forceNewSession: true,
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/task-agent-manager.ts]: Valid — confirmed against the code. The fresh post-approval session has no `node_executions` row: the reviewer row already has an `agentSessionId`, so `createSubSession`'s fresh path skips the write (`task-agent-manager.ts:1509-1512`, the `else if (match && match.agentSessionId)` warn branch), and `rehydrateSubSessionsForRun` discovers sessions **only** via `nodeExecutionRepo.listByWorkflowRun → execution.agentSessionId` (`:3274`). So a daemon restart inside the merge window would orphan the merge session and leave the task parked in `approved`.

This is real, but it's exactly the crash-window category task #850 deliberately deferred — its scope note states a reconciliation/recovery sweep is "optional future hardening ... not part of this task." This PR moves the merge from 0% (always interrupted) to runs-in-the-common-case; restart-window recovery is a separate follow-up. Not blocking; tracking as a known limitation. 👍 on the observation.

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts [hunk 2]
```diff
@@ -1261,7 +1272,12 @@ export class TaskAgentManager {
     // NodeExecution row with `agentSessionId` exists yet. Resolve the
     // eager session directly from the in-memory index so the reuse logic
     // below picks it up instead of creating a second session.
-    if (memberInfo?.agentName) {
+    //
+    // `forceNewSession` (post-approval spawn) bypasses reuse entirely: the
+    // target agent already has a live session that the approval/quiesce flow
+    // is interrupting, so reuse would inject the merge kickoff into a dying
+    // session. Fall straight through to the fresh-session path below.
+    if (memberInfo?.agentName && !memberInfo.forceNewSession) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/task-agent-manager.ts]: Valid edge case. The fresh path does call `flushPendingMessagesForTarget(runId, 'reviewer', sessionId)` (`task-agent-manager.ts:1547`), so any queued reviewer handoffs would drain into the new merge session instead of the original reviewer session.

In practice the reviewer has already drained its queue by the time it called `submit_for_approval`, so pending reviewer messages at post-approval time are a narrow race — and the merge kickoff (injected afterward at `spawnPostApprovalSubSession:4880`) carries explicit instructions that tolerate the noise. A `forceNewSession`-aware skip here (or a distinct queue target) is a reasonable hardening, but it sits outside #850's stated 3-line fix and the task's "surgical changes" guidance. Tracking as a follow-up rather than blocking the 100%-failure fix. 👍

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/task-agent-manager.ts]: Agreed — deferring per #850 scope. The fresh path inherits createSubSession write-once agentSessionId linking, so the merge session has no node_executions row; restart-during-merge recovery is the crash-window case the task scoped out. Strict improvement (0% → common case); the reconciliation sweep / rehydrate-from-postApprovalSessionId is the right follow-up. 'Send back' → re-approve is the interim manual recovery.

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/task-agent-manager.ts]: Agreed — deferring per the surgical-changes guidance. A forceNewSession-aware skip (or distinct queue target) in the fresh path is the clean follow-up; not blocking the 100%-failure fix.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (GLM) — Round 2

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** GLM

Recommendation: HOLD — push `4bf6b568b` before final approval (procedural, not a code defect).

**Daemon fix (`40df0384`, task #850):** unchanged from round 1, still approved. The two bot P2s are resolved (0 unresolved threads); both correctly deferred per #850's crash-window scope. No regression.

**UI rename (`4bf6b568b`):** verified locally — correct and clean.
- Label `Retry → Send back` is more truthful (the button resets `approved → in_progress`; `retryPostApproval` was never built) and matches the sibling `PendingTaskCompletionBanner` (`label: 'Send back'`, `:109`).
- No stale references to the old `pending-post-approval-retry-btn` testId / `onRetry` in the post-approval context. The other `Retry` hits (ConnectionOverlay, PendingGate/HookBanner, ErrorBoundary) are unrelated retry-failed-op buttons and correctly unchanged.
- 7/7 web tests pass; `oxlint` clean; `biome format` clean. (The 2 `biome check` `organizeImports` notices are not a project gate — lint is `oxlint`, biome is `format`-only, `assists: null` — and the imports are byte-identical on `dev`, so pre-existing.)

**Blocker:** `4bf6b568b` exists only locally — the PR head on GitHub is still `40df0384`, so CI has not run on the UI change and the PR does not contain the commit you're asking me to approve. Please push it to the branch, let CI go green on the new head, then I'll approve immediately. I don't want to approve a head that differs from the reviewed/described work.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (GLM) — Final (Round 3)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** GLM

Recommendation: APPROVE

Procedural blocker cleared — approving. PR head is now `4bf6b568b` (both commits pushed), CI fully green on the new head (all Daemon Unit shards incl. 5-space-runtime-a/b + 5-space-agent-other, the full Daemon Online matrix, Web Tests, Lint/Knip/Format/Typecheck, Coverage Gate; the 4 skipping checks only fire on push-to-dev), MERGEABLE, 0 unresolved threads.

Both commits reviewed in full over rounds 1–2:

**`40df0384` — #850 daemon fix.** `SubSessionMemberInfo.forceNewSession` bypasses the agentName reuse lookup (`task-agent-manager.ts:1280`), set only from `spawnPostApprovalSubSession` (`:4859`) — grep confirms exactly three production references, so normal node re-execution still reuses (hard constraint). Resolves the 100% post-approval merge failure: the spawn no longer hands back the reviewer's mid-teardown session, so it succeeds and `postApprovalSessionId` is stamped (`post-approval-router.ts:420`). No completion-callback regression (`mark_complete` transitions `approved→done` directly); quiesce exclusion protects the fresh id.

**`4bf6b568b` — UI rename.** `Retry → Send back` is truthful (the button resets `approved→in_progress`; `retryPostApproval` was never built) and matches `PendingTaskCompletionBanner:109`. No stale testId refs; 7/7 web tests pass; `oxlint` + `biome format` clean.

The two bot P2s (rehydration-after-restart, queue-drain pollution) are real but deferred per #850's explicit crash-window scope — both resolved on the PR, tracked as follow-ups.

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts [hunk 3]
```diff
@@ -4783,6 +4792,31 @@ export class TaskAgentManager {
...
+    // for workflows whose post-approval target is an agent that already ran.
+    const existingSessionId = this.findLiveSubSessionForAgent(task, matchedSlot.name);
+    if (existingSessionId) {
+      const existing = this.getSubSession(existingSessionId);
+      if (!existing) {
+        throw new Error(
+          `spawnPostApprovalSubSession: live session ${existingSessionId} for agent "${matchedSlot.name}" vanished before injection (task ${taskId})`
+        );
+      }
+      await this.injectMessageIntoSession(existing, kickoffMessage);
```
Comment (lsm) [inline packages/daemon/src/lib/space/runtime/task-agent-manager.ts:4812]: **P2 (follow-up, not blocking this PR)** — stale node-agent topology in the reuse branch.

This branch injects into the target agent's live session but deliberately skips `reinjectNodeAgentMcpServer` (to avoid `restartQuery()`). The reused session keeps the `node-agent` closure from its ORIGINAL spawn, whose `workflowNodeId` drives `send_message` topology/gate resolution. For a custom workflow whose `postApproval.targetAgent` is an agent that already ran on a *different* node than `matchedNodeId`, `send_message` from the merge turn would evaluate channels from the old node and could fail/route oddly.

Scope is narrow: the built-in `merger` target creates fresh (unaffected), and `mark_complete` (the actual approved→done path) is `workflowNodeId`-agnostic, so the typical post-approval flow is fine. But if a custom workflow's post-approval kickoff directs the reused agent to message peers, consider a no-restart node-agent rebuild here (or document the constraint). Tracking as a follow-up — not blocking the #816 fix.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (GLM) — Round 4 (design reversal)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** GLM

Recommendation: APPROVE

Reviewed the new reuse-if-exists-else-create design from scratch (this needed a real review, not just CI). The reversal is sound and the new diagnosis is correct.

**Diagnosis correction confirmed.** Task #850's premise was wrong: the built-in Coding workflow targets `merger` (`built-in-workflows.ts:546`, slot at `:559`), not `reviewer` — so the reviewer-session-reuse story never applied to the built-in path. The real failure mechanism is `createSubSession`'s reuse path → `reinjectNodeAgentMcpServer` → `session.restartQuery()` (`task-agent-manager.ts:4147/4157`), which interrupts an *active* SDK drive ("Interrupted by user"). Verified `restartQuery` is a no-op only when idle, and genuinely restarts/interrupts when a query is running.

**Built-in `merger` path is clean.** `findLiveSubSessionForAgent('merger')` returns null (the Post-Approval node is never activated pre-approval → no `node_executions` row), so the code takes the create branch → `createSubSession` creates genuinely fresh → no `restartQuery`, no interruption. (Side note, pre-existing/unchanged: the entire `eagerSubSessionIds` / `eagerlySpawnWorkflowNodeAgents` mechanism is dead code — the map is never populated. Worth a separate cleanup, not this PR.)

**Reuse branch is safe.** Verified `injectMessageIntoSession` enqueues when busy and starts a fresh turn only when idle — it never aborts/restarts/interrupts an active drive (`message-queue.ts:160`, `query-lifecycle-manager.ts:615-620`). So the reuse path for custom workflows whose `targetAgent` already ran no longer hits the interrupt. `forceNewSession` is cleanly removed (grep: zero refs in src/tests); the `createSubSession` guard is back to `if (memberInfo?.agentName)`, so normal node re-execution still reuses — hard constraint preserved.

**Tests** cover the decision boundary (reuse live + inject directly; normal reuse on 2nd activation; create when no prior). CI fully green on `146e219c2` (32 checks, 0 fail, 4 expected dev-only skips); MERGEABLE; 0 unresolved threads.

**Documented follow-ups (non-blocking):**
- **P2** (inline comment on the reuse branch): the reused session keeps its original `workflowNodeId` for `send_message` topology. Narrow — only custom workflows whose targetAgent ran on a different node AND whose merge directs peer messaging; `mark_complete` (the approved→done path) is node-agnostic and the built-in path is unaffected.
- **P3** (cosmetic): the reuse branch skips `registerCompletionCallback`, so the reused session's `node_executions` row stays `in_progress`. Doesn't affect task status (merge completes via `mark_complete`).

The earlier two bot P2s (rehydration-after-restart, flush pollution) remain deferred per #850 scope and are unchanged by this design swap.

File: packages/daemon/src/lib/space/runtime/task-agent-manager.ts (hunk 3 above)
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/task-agent-manager.ts:4812]: Valid but narrower than stated — concur it is a follow-up, not blocking. In the common case there is no mismatch: matchedNodeId is the node declaring the targetAgent slot, and the reused session node-agent closure was built at that same node (an agent runs where it is declared), so send_message topology lines up. The stale-closure risk only arises when the targetAgent slot is declared on a different node than the one the live session ran on (multi-declaration or agentId cross-match) AND the kickoff directs the reused agent to message peers. Built-in merger is unaffected (fresh session); mark_complete is workflowNodeId-agnostic, so the typical approved-to-done path is fine. Right follow-up: a no-restart node-agent rebuild here (merge the matchedNodeId closure without restartQuery, so the kickoff turn picks up the fresh topology) — tracking separately.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (GLM)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** GLM

**Recommendation: REQUEST_CHANGES**

Round-5 review of the per-node fan-out redesign (head `cf911a79d`). CI green (all 13 runnable jobs), mergeable/clean, all prior review threads resolved.

## What's correct (verified independently)

- **#850/#816 fix is sound.** The built-in single-route `merger` path is atomic: `spawnPostApprovalSubSession` reuses the target agent's live session via direct injection (`injectMessageIntoSession` — never interrupts the SDK drive, the real #816 cause) or creates one fresh; the `postApprovalSessionId` stamp either lands or the throw propagates with no half-state. `createSubSession`'s reuse guard is back to its original `agentName` form — the one-session-per-agent-per-task constraint is intact.
- **Migration is safe.** `mergeNodeStructuralFieldsFromTemplate:1437` overwrites node `postApproval` from the template by name. Template Review/QA nodes carry no route (clears any stale route on existing spaces); the Post-Approval node carries the `merger` route. No built-in can end up with two routes. Confirmed no other writer (m94 backfill, `sanitizePostApprovalForLoad`, preset-agent backfill) re-adds a route to the Review node.
- **Callers are throw-safe.** All 3 entry points (`approvePendingCompletion` RPC, tick-loop end-node path, canonical-task resolution) transition to `approved` before routing, catch dispatch throws via `mapPostApprovalDispatchWarning` so approval is never lost, and don't re-invoke on success. The `finally` pending-completion cleanup holds.
- **No double-dispatch via legacy.** `collectPostApprovalRoutes` returns node routes OR the legacy workflow-level route, never both.

## P2 — multi-route fan-out partial-failure (new code in this PR)

The new fan-out loop has no per-route error handling. `post-approval-router.ts:402-445`: if route 1 spawns (primary set at :424) and route 2's `spawnPostApprovalSubSession` throws, the `postApprovalSessionId` stamp at :437 never runs, leaving it null. On retry/re-dispatch the double-fire guard (:375) reads null and re-dispatches **all** routes — double-delivering the kickoff to route 1's already-running session.

This is **not reachable for any built-in workflow** (Coding/Research/QA each declare exactly one route → the loop is atomic). It only affects a custom workflow with ≥2 nodes carrying `postApproval` routes. Even then it's recoverable (a merger re-dispatch is largely idempotent), but it's a real double-fire bug in new code this PR ships, and there's no test covering "route 1 succeeds, route 2 throws."

Two cheap fixes, pick one:
1. **Stamp-on-first-success + per-route try/catch:** stamp `postApprovalSessionId` (or track a set) as soon as the first route spawns, and catch each route's throw so remaining routes still dispatch and the primary is always tracked. ~5 lines.
2. **Validator cap:** have `post-approval-validator` reject workflows declaring >1 `postApproval` route until the fan-out is hardened. Keeps the loop simple.

Given Commit 2 (move `postApproval` to agent-level, drop `targetAgent`) is deferred and will reshape this loop anyway, option 1 is the smaller interim patch.

## P3 — nit (optional)

`built-in-workflows.test.ts` has no test asserting the specific migration case "an existing Review node carried a stale `postApproval` route → restamp clears it." The code at `:1437` handles it (template Review's `undefined` overwrites), but the existing restamp test (`:1690`) only simulates the Post-Approval route being cleared, not a stale Review route. A one-line assertion would lock the invariant.

## Summary

The core fix is correct and ready; the only blocker is hardening the new multi-route fan-out loop (P2). The 3 previously-deferred P2 follow-ups (rehydration on restart, handoff-queue drain, stale node-agent topology in the reuse branch) remain non-blocking as noted.

File: packages/daemon/src/lib/space/runtime/post-approval-router.ts [hunk 4]
```diff
@@ -379,55 +387,69 @@ export class PostApprovalRouter {
...
+        continue;
+      }
+      const kickoffMessage = appendPostApprovalCompletionInstructions(interpolatedInstructions);
+      const { sessionId } = await this.deps.spawner.spawnPostApprovalSubSession({
+        task,
+        workflow,
+        targetAgent: route.targetAgent!,
+        kickoffMessage,
+      });
+      if (!primarySessionId) primarySessionId = sessionId;
```
Comment (lsm) [inline packages/daemon/src/lib/space/runtime/post-approval-router.ts]: **P2 — partial-failure gap in the new fan-out loop.** `primarySessionId` is tracked here, but `postApprovalSessionId` is only stamped AFTER the loop completes (`:437`). If a later route's `spawnPostApprovalSubSession` throws, the stamp never runs → on retry the double-fire guard (`:375`) sees a null id and re-dispatches **every** route, double-delivering the kickoff to this route's already-running session.

Not reachable for built-in (single-route) workflows — those stay atomic. It only bites a custom workflow with ≥2 `postApproval` routes, and there is no test for "route 1 succeeds, route 2 throws."

Cheapest fix: stamp `postApprovalSessionId` right here on first success (or track a set of spawned ids) AND wrap each spawn in try/catch so a later-route failure cannot un-track the primary. Alternative: have `post-approval-validator` cap postApproval routes at one until the fan-out hardens. See review body for full context.

Comment (lsm) [reply packages/daemon/src/lib/space/runtime/post-approval-router.ts]: Fixed in 792cbff08. I went a step beyond option 1 toward the spirit of your option 2: the router still scans every node (so the merger fires regardless of which node submitted/approved), but it now dispatches AT MOST ONE route — the first declared, warning on extras. The multi-spawn loop is gone entirely, so there's no partial-failure window: a single spawn, stamped immediately on success, and the double-fire guard sees a non-null id on retry.

Why this over stamp-on-first-success + try/catch: option 1 would have left the related multi-route gaps the bot flagged still open (mark_complete race, sibling-quiesce only excluding the primary, and the renamed-node migration double-injection). Collapsing to single dispatch fixes the P2 partial-failure and defuses those at their root, with no validator/schema/reconciler changes. Commit 2 moves the route to the agent slot and reshapes this further, so heavy loop hardening now would be wasted.

Tests: post-approval-router now asserts single dispatch on multi-route input (first route wins); built-in-workflows adds the P3 restamp test you flagged (stale Review-node route cleared on re-template → exactly one merge route). All 2954 5-space daemon tests + tsc/oxlint/biome/knip green.

One bot thread I'm explicitly NOT fixing here: the renamed-Post-Approval-node case (route vanishes → approval skips merge) is a real but orthogonal migration-completeness edge case — I've replied on that thread with the plan and propose it as a follow-up so this PR stays on the partial-failure blocker. Happy to pull it in if you'd rather.

File: packages/daemon/src/lib/space/runtime/post-approval-router.ts [hunk 5]
```diff
@@ -379,55 +387,69 @@ export class PostApprovalRouter {
...
+        );
+        continue;
+      }
+      const kickoffMessage = appendPostApprovalCompletionInstructions(interpolatedInstructions);
+      const { sessionId } = await this.deps.spawner.spawnPostApprovalSubSession({
+        task,
+        workflow,
+        targetAgent: route.targetAgent!,
+        kickoffMessage,
+      });
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/post-approval-router.ts]: Fixed in 792cbff08 — see the full write-up on the review thread (discussion_r3725547669). The fan-out loop is replaced with single-route dispatch: at most one spawn, stamped immediately on success. There's no longer a window where an earlier route is running untracked after a later throw, and the double-fire guard sees a non-null id on retry.

[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/post-approval-router.ts]: Resolved by 792cbff08 (single-route dispatch): only one route receives the mark_complete instruction, so there's no first-finisher race.

Separately, even before this change a second mark_complete was a benign soft no-op — end-node-handlers.ts:147-154 returns success:false ("task is not in approved") without throwing or mutating state, so a slower worker would just stop. But single-route dispatch makes the question moot: there's only ever one completer.

File: packages/daemon/src/lib/space/runtime/post-approval-router.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/post-approval-router.ts]: Resolved by 792cbff08 (single-route dispatch): there is now at most one post-approval session, which the existing exclusion at space-runtime.ts:8014 (e.agentSessionId !== spawnedPostApprovalSessionId) fully covers.

The original concern only arose with two routes targeting different agents, which no built-in or migration produces (the renamed-node migration case yields two routes both targeting 'merger' — one session).

File: packages/daemon/src/lib/space/runtime/post-approval-router.ts [hunk 6]
```diff
@@ -210,23 +210,23 @@ export function appendPostApprovalCompletionInstructions(interpolatedInstruction
...
+ * workflow-level route as a fallback) regardless of which node submitted or
+ * approved. Each collected route is delivered to its own `targetAgent` session.
+ */
+function collectPostApprovalRoutes(workflow: SpaceWorkflow | null): PostApprovalRoute[] {
+  if (!workflow) return [];
+  // Approval is a task-level event: collect post-approval routes from EVERY
+  // node, independent of which node submitted or approved.
+  const nodeRoutes = workflow.nodes
+    .map((node) => node.postApproval)
+    .filter((route): route is PostApprovalRoute => !!route);
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/post-approval-router.ts:227]: Addressed by 792cbff08. For the normal (non-renamed) case, mergeNodeStructuralFieldsFromTemplate:1437 clears the stale Review-node route and stamps Post-Approval — now covered by a new restamp test asserting exactly one merge route survives.

For the renamed-Review edge case you describe (stale route preserved on the unmatched node + new route on Post-Approval → two routes both targeting 'merger'): single-route dispatch now delivers only the first, so there's no double dispatch / concurrent merge attempt. Both routes target 'merger' with identical merge instructions, so the first carries the correct kickoff.

File: packages/daemon/src/lib/space/workflows/built-in-workflows.ts [hunk 7]
```diff
@@ -539,27 +539,25 @@ export const CODING_WORKFLOW: SpaceWorkflow = {
...
         {
           agentId: 'PR Merger',
           name: 'merger',
           customPrompt: PR_MERGER_SLOT_PROMPT,
         },
       ],
+      postApproval: {
+        targetAgent: 'merger',
+        instructions: PR_MERGE_POST_APPROVAL_INSTRUCTIONS,
+      },
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/workflows/built-in-workflows.ts:560]: Real edge case, but orthogonal to the multi-route concerns this commit addresses — proposing it as a focused follow-up rather than expanding this PR.

Confirmed the mechanism you describe: if a user renamed the Post-Approval node, missingTemplateNodes won't re-append it (the merger agent already exists in the renamed node), and the renamed node carries no route (pre-redesign the route lived on Review), so approval transitions to done without merging. Pre-PR this was harmless (route was on Review); the route-move makes it a regression for that narrow renamed-node config.

The proportionate fix lives in the reconciler — match the route-owning template node by its merger slot rather than only by node name, so the route lands on the renamed node. That's separate reconciler surgery from the partial-failure blocker this PR is scoped to, and Commit 2 (agent-level move) reshapes the model anyway. Flagging for visibility; happy to do it here if you'd prefer it not wait.
